compositor_pipeline/src/pipeline/input/mp4.rs

```diff
@@ -0,0 +1,269 @@
+use std::{
+    fs::File,
+    io::Read,
+    os::unix::fs::MetadataExt,
+    path::PathBuf,
+    sync::{atomic::AtomicBool, Arc},
+    thread::JoinHandle,
+    time::{Duration, Instant},
+};
+
+use bytes::{Buf, Bytes, BytesMut};
+use compositor_render::InputId;
+use crossbeam_channel::Receiver;
+use mp4::Mp4Reader;
+
+use crate::pipeline::{
+    structs::{EncodedChunk, EncodedChunkKind},
+    VideoCodec,
+};
+
+use super::ChunksReceiver;
+
+pub struct Mp4Options {
+    pub source: Source,
+    pub input_id: InputId,
+}
+
+pub enum Source {
+    Url(String),
+    File(PathBuf),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum Mp4Error {
+    #[error("Error while doing file operations.")]
+    IoError(#[from] std::io::Error),
+
+    #[error("Error while downloading the MP4 file.")]
+    HttpError(#[from] reqwest::Error),
+
+    #[error("Mp4 reader error.")]
+    Mp4ReaderError(#[from] mp4::Error),
+
+    #[error("No suitable track in the mp4 file")]
+    NoTrack,
+}
+
+pub struct Mp4 {
+    pub input_id: InputId,
+    video_thread: Option<std::thread::JoinHandle<()>>,
+    delete_file: bool,
+    path: PathBuf,
+    stop_video_reader: Arc<AtomicBool>,
+}
+
+impl Mp4 {
+    pub fn new(options: Mp4Options) -> Result<(Self, ChunksReceiver), Mp4Error> {
+        let (input_path, delete_file) = match options.source {
+            Source::Url(url) => {
+                let mut file_response = reqwest::blocking::get(url)?;
```

Comment:
check the status code, unless I'm mistaken get will return error only if request failes, but if it returns 4xx or 5xx it will not.